exercises/dominik_arominski/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,154 @@
+#include <iostream>
+#include <string>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+class Shape {
+
+protected:
+  std::string type;
+  int nSides;
+
+public:
+  
+  std::string GetType() const { return type;};
+  int GetSides() const { return nSides;};
+  
+  virtual double CalculatePerimeter() const { return 0;};  
+  virtual double CalculateArea() const { return 0;};  
+
+  friend std::ostream& operator<<(std::ostream& out, const Shape& shape) {
+    return out << shape.GetType() << "\t" << shape.GetSides() << "\t" << shape.CalculateArea() << "\t" << shape.CalculatePerimeter();
+  }
+  
+};
+
+class Triangle : public virtual Shape {
+private:
+  double height;
+  double base;
+public:
+  Triangle(double fHeight, double fBase) : height(fHeight), base(fBase) {
+    type = "Triangle";
+    nSides = 3;
+  }
+  
+  double CalculatePerimeter() const {
+    return base + 2.0 * sqrt(pow(height, 2) + 0.25 * pow(base, 2));
+  }
+  double CalculateArea() const {
+    return 0.5 * base * height;
+  }
+
+};
+
+class Circle : public virtual Shape {
+private:
+  double radius;
+public:
+  Circle(double fRadius=0) : radius(fRadius) {
+    type = "Circle";
+    nSides = 0;
+  }
+
+  double CalculatePerimeter() const {
+    return 2.0 * M_PI * radius;
+  }
+  double CalculateArea() const {
+    return M_PI * pow(radius, 2);
+  }
+};
+
+class Rectangle : public virtual Shape {
+protected:
+  double side1;
+  double side2;
+public:
+  Rectangle(double fSide1=0, double fSide2=0) : side1(fSide1), side2(fSide2) {
+    type = "Rectangle";
+    nSides = 4;
+  }
+
+  double CalculatePerimeter() const {
+    return 2 * side1 + 2 * side2;
+  }
+  double CalculateArea() const {
+    return side1 * side2;
+  }
+  
+};
+
+class Square : public Rectangle {
+public:
+  Square(double fSide=0) : Rectangle(fSide) {
+    side2 = side1;
+    type = "Square";
+  }
+};
+
+class ShapeSorter {
+
+public:
```

Comment:
For these methods, it would be nicer to separate the printing from the sorting/filtering. Whether the sorting should order the original vector, or make a copy, is not specified in the exercise, so you can decide.